src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1119,50 +1057,82 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
+                                          destinationId.size()));
 
-        if (mOpCredSet->IsTrustedRootIn(trustedRoot))
-        {
-            if (!mTrustedRootId.empty())
-            {
-                chip::Platform::MemoryFree(const_cast<uint8_t *>(mTrustedRootId.data()));
-                mTrustedRootId = CertificateKeyId();
-            }
-            mTrustedRootId = CertificateKeyId(reinterpret_cast<const uint8_t *>(chip::Platform::MemoryAlloc(kTrustedRootIdSize)));
-            VerifyOrReturnError(!mTrustedRootId.empty(), CHIP_ERROR_NO_MEMORY);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASESession::FindDestinationIdCandidate(const ByteSpan & destinationId, const ByteSpan & initiatorRandom)
+{
+    uint8_t nCertificateSets = mOpCredSet->GetCertCount();
+
+    for (uint8_t i = 0; i < nCertificateSets; ++i)
+    {
+        uint8_t candidate[kSHA256_Hash_Length] = { 0 };
+        CertificateKeyId trustedRootId;
+        ChipCertificateData nodeOperationalCertificate;
+        NodeId nodeId;
+        FabricId fabricId;
+
+        trustedRootId = mOpCredSet->GetTrustedRootId(i);
+
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(trustedRootId), mOpCredSet->GetDevOpCredLen(trustedRootId),
+                                            nodeOperationalCertificate));
+
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertChipId(nodeId));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
 
-            memcpy(const_cast<uint8_t *>(mTrustedRootId.data()), trustedRoot.data(), trustedRoot.size());
+        {
+            MutableByteSpan candidateSpan(candidate, sizeof(candidate));
+            ReturnErrorOnFailure(GenerateDestinationID(initiatorRandom, &trustedRootId, nodeId, fabricId, candidateSpan));
+        }
 
+        if (memcmp(destinationId.data(), candidate, destinationId.size()) == 0)
+        {
+            mTrustedRootId = trustedRootId;
             break;
         }
     }
+
     VerifyOrReturnError(!mTrustedRootId.empty(), CHIP_ERROR_CERT_NOT_TRUSTED);
 
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CASESession::ConstructSaltSigmaR2(const ByteSpan & rand, const P256PublicKey & pubkey, const uint8_t * ipk,
-                                             size_t ipkLen, MutableByteSpan & salt)
+CHIP_ERROR CASESession::ConstructSaltSigmaR2(const ByteSpan & rand, const Crypto::P256PublicKey & pubkey, const ByteSpan & ipk,
+                                             MutableByteSpan & salt)
 {
     uint8_t md[kSHA256_Hash_Length];
     memset(salt.data(), 0, salt.size());
     Encoding::LittleEndian::BufferWriter bbuf(salt.data(), salt.size());
 
-    bbuf.Put(ipk, ipkLen);
+    bbuf.Put(ipk.data(), ipk.size());
     bbuf.Put(rand.data(), kSigmaParamRandomNumberSize);
     bbuf.Put(pubkey, pubkey.Length());
     ReturnErrorOnFailure(mCommissioningHash.Finish(md));
```

Comment:
The Finish() followed by Begin() on mCommissioningHash means that we are losing the "total" transcript. The transcript was meant to be only Msg1:

![image](https://user-images.githubusercontent.com/1313847/126172419-930417a9-6887-4728-ae5f-e2206336e8d8.png)

The full transcript hash is required to generate the session key, but it will be lost from this new Begin().

See:
![image](https://user-images.githubusercontent.com/1313847/126181782-e6109f98-f320-4b3a-bc1d-a6f0249b7367.png)


@balducci-apple Could you please take a peek at transcript computation overall in this branch?